pkg/parser/parser.go

```diff
@@ -26,22 +26,36 @@
 package parser
 
 import (
+	"fmt"
 	"strings"
+
+	"github.com/raklaptudirm/mash/pkg/command"
+	"github.com/raklaptudirm/mash/pkg/command/builtin"
 )
 
+var NoCommandError = fmt.Errorf("No commands provided")
+
 // Function Parse accepts user command as
 // input and separates the command string
 // from the argument string, and then
 // dispatches the argument string to the
 // Args function. It returns the command
 // and the argument slice.
-func Parse(input string) (string, []string) {
+func Parse(input string) (command.Command, error) {
 	input = strings.Trim(input, " \t\n\r")
 	words := Words(input)
 	if len(words) == 0 {
-		return "", []string{}
+		return command.Command{}, NoCommandError
+	}
+	cmd, exists := builtin.Commands[words[0]]
+	if !exists {
+		return command.Command{
+			Name: words[0],
+			Args: words[1:],
+		}, nil
 	}
-	return words[0], words[1:]
+	cmd.Args = words[1:]
+	return cmd, nil
```

Comment:
If we have a command map inside the shell struct then this would be very easy to check that the command doesn't exist and since the cmd.Execute is already checking whether this is a builtin command then we don't need to do anything other than just check that this command exists and pass it along. If it doesnt exist then it can return an error.